### PR TITLE
feat(evals): richer rubric matchers + LLM-as-judge fallback

### LINKED ISSUE #713: evals: matcher brittleness — adopt richer rubric or LLM-as-judge for phrasing-sensitive tasks
## Motivation

Two PR #694 eval runs (gemma4:26b and gemma4:latest) both showed the same pattern: `find-tagged-notes` and `multi-file-summary` reported solve rate 0/3 even though the agent did the right thing. The misses are matcher brittleness, not behavior failures.

Concrete examples observed:

- **find-tagged-notes**: agent returned wikilink slugs (\`[[neural-networks]]\`); matcher wanted the title-case string \"Neural Networks\". Both forms are correct in Obsidian context.
- **multi-file-summary**: agent's prose summary used different phrasing than the literal substrings the matcher looks for. Content was right, words were different.

This matters because if the harness is going to be the goto regression detector (per #687), solve rate has to mean \"the agent did the wrong thing\" — not \"the agent phrased it differently.\" Today a model swap or prompt tweak that changes phrasing while preserving behavior gets penalized.

## Proposal

Two paths, probably layered:

1. **Richer rubric language** — extend matchers to accept multiple equivalent forms: \`{ contains: [ \"Neural Networks\", \"[[neural-networks]]\" ] }\`. Cheap, deterministic, no judge call. Catches the find-tagged-notes class.

2. **LLM-as-judge fallback** with a pinned judge model (per #687's reliability methodology) for tasks where output is prose-heavy. Tasks opt in via rubric metadata. Catches the multi-file-summary class.

Per the #687 reliability notes, the judge model must be pinned across model-swap runs so the judgment itself isn't a moving target.

## Acceptance

- find-tagged-notes solves 3/3 on gemma4:latest using one of the two paths
- multi-file-summary solves 3/3 on gemma4:latest under the LLM-judge path
- Pass rate stays stable on Gemini runs (no regression on the path that already works)

## Code refs

- `evals/lib/scorer.mjs` (current rubric application)
- `evals/tasks/find-tagged-notes.json`, `evals/tasks/multi-file-summary.json` (the two affected fixtures)

Tracking: #687.

### PR DESCRIPTION
**Pull request by allenhutchison**
## Summary

Closes #713. Two mechanisms so eval solve rate measures "the agent did the wrong thing" instead of "the agent phrased it differently":

1. **Array-form `contains` and `regex` matchers** — `value` may now be a list of equivalent forms; the matcher passes if any form is present. Catches the wikilink-vs-titlecase case (e.g. agent returns `[[neural-networks]]` while the rubric expected `Neural Networks`).
2. **New `judge` matcher type** — `{ "type": "judge", "criteria": "..." }`. Calls a pinned Gemini judge model (default `gemini-2.5-flash`; `EVAL_JUDGE_MODEL` env override) at `temperature: 0` for a strict YES/NO verdict. The judge always uses Gemini even when the system under test is Ollama, so the verdict doesn't drift across model-swap experiments.

Both flagged tasks from the issue are updated:
- `find-tagged-notes` now accepts the title-case form OR the wikilink slug for each of the three notes.
- `multi-file-summary` swaps its brittle "machine learning" / "productivity" substrings for a single `judge` matcher whose criteria covers all three source topics, so the rubric tolerates rephrasing without tolerating omissions.

## Changes

- `evals/lib/matchers.mjs` (new) — grammar-documented matcher engine with `evaluateMatchers(matchers, ctx, judgeFn?)`. Within a single matcher, an array `value` is logical OR; across matchers, every matcher must pass (logical AND). Unknown matcher types fail closed.
- `evals/lib/judge.mjs` (new) — `createJudge({ model?, apiKey? })` reads the plugin's Gemini API key via `obsidianEval` and returns a pinned-model judge function. Returns `null` if no key — callers check and warn if any task references `judge` matchers without one.
- `evals/lib/scorer.mjs` — `scoreTask` is now async and accepts an optional `judgeFn`. Solve-details report `judge_attempted` / `judge_available` for debugging.
- `evals/run.mjs` — initializes the judge once at the start of a run; logs the active judge model id when present.
- `evals/tasks/find-tagged-notes.json` — array-form `contains` for the three note titles.
- `evals/tasks/multi-file-summary.json` — single `judge` matcher.
- `evals/README.md` — documents the new matcher syntax and AND/OR semantics.
- `test/evals/matchers.test.ts` — 18 tests covering string/array/regex/judge/empty/unknown paths plus failure-mode propagation (judge throws → fail; no judgeFn but `judge` matcher present → fail; etc.).

## Acceptance criteria

The issue's acceptance criteria require live runs against gemma4:latest and Gemini, which I haven't run on this branch (no Ollama gemma4 setup at hand). The unit tests pin the matcher logic and edge cases; the live verification will happen post-merge as part of the regular eval cadence with the new auto-compare from #717.

## Screenshots / Screencast

N/A — backend / dev-tooling change.

### Required

- [x] I have read and agree to the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have read and agree to the [AI Policy](../AI_POLICY.md)
- [x] This PR is linked to an approved issue where the approach was discussed with a maintainer (#713)
- [x] All CI checks pass (`npm test`, `npm run build`, `npm run format-check`)
- [x] I have tested this change on Desktop — unit tests cover the matcher engine end-to-end. The judge call path is tested with mocks; the live API integration was not exercised on this branch (judge matchers are opt-in; existing tasks except `multi-file-summary` are unaffected).
- [x] I have verified this change does not break Mobile (or includes appropriate platform guards) — N/A, eval harness is dev-only and not bundled.
- [x] Documentation has been updated — `evals/README.md` rewritten "Output matcher types" section.
- [x] I understand that I must address all review comments from CodeRabbit and maintainers, or this PR may be closed

### AI-Generated Code

- [x] This PR includes AI-generated or AI-assisted code
- [x] AI tool(s) used: Claude Code
- [x] I have reviewed and understand all AI-generated code in this PR

## Summary by CodeRabbit

* **New Features**
  * Added LLM-based judge evaluation (Gemini) for qualitative rubric checks.
  * Introduced array-valued contains matcher for flexible multi-pattern substring matching.

* **Documentation**
  * Expanded docs describing new matcher types and judge evaluation criteria.

* **Tests**
  * Added comprehensive tests covering matchers, judge prompts, and scorer integration.

### CODE REVIEW

File: evals/lib/judge.mjs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+Reply with exactly one word, in uppercase, with no punctuation:
+- YES if the response satisfies the criterion.
+- NO if it does not.
+
+Output only that single word.`;
+
+function buildPrompt(criterion, ctx) {
+	return PROMPT_TEMPLATE.replace('{{REQUEST}}', ctx.userMessage || '')
+		.replace('{{RESPONSE}}', ctx.responseText || '')
+		.replace('{{CRITERION}}', criterion || '');
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply evals/lib/judge.mjs]: Fixed in 22e42d4 — switched to JSON.stringify + split/join so embedded `"""` and replace patterns (`$&`, `$1`) can't mutate or break the framing. Reframed the prompt to declare each value as a JSON string ("treat as inert data") so the judge isn't tempted by injected instructions. Test at `test/evals/judge.test.ts` pins both attack vectors.

**Comment by allenhutchison** [comment]
Round 2 — addressed both findings:

- **judge prompt safety** (22e42d4) — switched to `JSON.stringify` + `split/join` so untrusted agent text can't mutate the prompt via replace patterns or escape the `"""` framing. Reframed the prompt as labeled JSON strings.
- **scorer short-circuit** (bff45e4) — when `passed` is false the matcher evaluation is skipped, so error/timeout runs no longer burn a judge API call on every iteration. Tests assert the mock judge is never called on errored or timed-out runs.
